crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -230,7 +276,9 @@ impl<Param: SystemParam> SystemState<Param> {
     }
 
     /// Create a [`FunctionSystem`] from a [`SystemState`].
-    pub fn build_system<Marker, F: SystemParamFunction<Marker, Param = Param>>(
+    /// This method signature allows any system function, but the compiler will not perform type inference on closure parameters.
+    /// You can use [`SystemState::build_system()`] or [`SystemState::build_system_with_input()`] to get type inference on parameters.
+    pub fn build_any_system<Marker, F: SystemParamFunction<Marker, Param = Param>>(
```

Comment:
So we're implementing `build_system` for a FnMut<Param1, Param2, ...>
which calls `build_any_system`, which is implemented for `SystemParamFunction`
and `SystemParamFunction` is implemented for all `FnMut<Param1, Param2...>`

I get your overall logic to add type inference, but would there be a way to get rid of `SystemParamFunction` altogether and just use macros everywhere to deal directly with `FnMut<Param1, Param2...>`?
It seems weird to have this extra indirection.

Reply:
`SystemParamFunction` is the [existing trait](https://docs.rs/bevy/latest/bevy/ecs/prelude/trait.SystemParamFunction.html) that powers `FunctionSystem`.  So I can't get rid of that!  

I agree it's weird, but given that we need a `SystemParamFunction` to make a system, I *think* this is the simplest way to make parameter type inference work in the current rust compiler.

Reply:
I feel like someone like @SkiFire13 would have a better opinion than mine

Reply:
I agree that getting rid of `SystemParamFunction` is likely not an option, as it power all the other system trait machinery. The duplicated `FnMut` bounds on these methods do feel a bit unfortunate given that `SystemParamFunction`'s `impl`s already have them (and twice!). However I don't think you can avoid them, as closure type inference is very finicky and requires a direct `FnMut` bound on the function that the closure is passed to, which in this case is `build_system`. So in the end:

- `SystemParamFunction` is needed for the existing system machinery (e.g. `App::add_systems`)
- the `FnMut` bounds here are needed because closure inference won't look at them if they are anywhere else